tools/cli/helpers/doc-parser/src/class-doc-parser.php

```diff
@@ -104,15 +146,304 @@ function ( $item ) use ( $ignore ) {
+
+					if ( $first && false === strpos( $line, '*/' ) ) {
+
+						break;
+					} else {
+						$first = false;
+					}
+
+					array_unshift( $docblock, $line );
+					if ( false !== strpos( $line, '/**' ) ) {
+						break;
+					}
+				}
+				$docblock = implode( '', $docblock );
+
+				$block['doc']                     = array();
+				$block['doc']['description']      = '';
+				$block['doc']['long_description'] = '';
+				$block['doc']['tags']             = array();
+
+				try {
+					$tokens     = new TokenIterator( $this->lexer->tokenize( $docblock ) );
+					$phpDocNode = $this->pdparser->parse( $tokens );
+				} catch ( ParserException $e ) {
+					continue;
+				}
+
+				foreach ( $phpDocNode->children as $entry ) {
+					if ( ! $entry instanceof PhpDocTextNode ) {
+						continue;
+					}
+
+					if ( ! empty( $entry->text ) ) {
+						$block['doc']['description'] .=
+							'<p>'
+							. str_replace( array( "\r\n", "\n", "\r" ), '</p><p>', $entry->text )
+							. '</p>';
+					}
+				}
+
+				$paramTags  = $this->get_param_tag_nodes( $phpDocNode );
+				$parameters = array();
+
+				foreach ( $paramTags as $paramTag ) {
+					$block['doc']['tags'][] = array(
+						'name'     => 'param',
+						'content'  => $paramTag->description,
+						'types'    => array(
+							(string) $paramTag->type,
+						),
+						'variable' => $paramTag->parameterName,
+					);
+
+					$parameters[] = (string) $paramTag . PHP_EOL;
+				}
+
+				foreach (
+					array(
+						'@since',
+						'@module',
+						'@deprecated',
+						'@see',
+						'@uses',
+						'@link',
+						'@type',
+					) as $tagType
+				) {
+					$sinceTags = $phpDocNode->getTagsByName( $tagType );
+					foreach ( $sinceTags as $sinceTag ) {
+						$block['doc']['tags'][] = array(
+							'name'    => substr( $tagType, 1 ),
+							'content' => (string) $sinceTag->value,
+						);
+					}
+				}
+			}
+
+			$filepath = ltrim( substr( $file, strlen( $path ) ), DIRECTORY_SEPARATOR );
+			$blocks[] = array(
+				'path'    => $filepath,
+				'root'    => $path,
+				'classes' => array(
+					array(
+						'methods' => array(
+							array(
+								'hooks' => $file_blocks,
+							),
+						),
+					),
+				),
+			);
+		}
 
 		if ( 'json' === $format ) {
-			$output = json_encode( $output, JSON_PRETTY_PRINT );
+			$blocks = json_encode( $blocks, JSON_PRETTY_PRINT );
+		}
+
+		return $blocks;
+	}
+
+	/**
+	 * Returns all param tag nodes. Tries to recover any parsing errors because of invalid markup.
+	 *
+	 * @param PhpDocNode $phpdocnode the parsed PHPDoc node.
+	 * @return array an array of PhpDocTagNode objects.
+	 */
+	public function get_param_tag_nodes( PhpDocNode $phpdocnode ): array {
+		$tags = $phpdocnode->getParamTagValues();
+
+		// Looking for invalid param tags.
+		foreach ( $phpdocnode->getTags() as $tag ) {
+			'@phan-var \PHPStan\PhpDocParser\Ast\PhpDoc\PhpDocTagNode $tag';
+
+			if ( $tag->name === '@param' && $tag->value instanceof InvalidTagValueNode ) {
```

Comment:
A useful `@todo` might be to have a flag to output diagnostics about this situation, which we could use in a CI check.